drivers/dma/dma_silabs_ldma.c

```diff
@@ -0,0 +1,564 @@
+		break;
+	case PERIPHERAL_TO_PERIPHERAL:
+	case HOST_TO_MEMORY:
+	case MEMORY_TO_HOST:
+	default:
+		return -ENOTSUP;
+	}
+
+	xfer_config->ldmaReqDis = false;
+	xfer_config->ldmaDbgHalt = false;
+
+	if (!dma_silabs_get_priority(config->channel_priority, &xfer_config->ldmaCfgArbSlots)) {
+		return -EINVAL;
+	}
+
+	switch (config->head_block->source_addr_adj) {
+	case DMA_ADDR_ADJ_INCREMENT:
+		xfer_config->ldmaCfgDstIncSign = ldmaCfgSrcIncSignPos;
+		break;
+	case DMA_ADDR_ADJ_DECREMENT:
+		xfer_config->ldmaCfgDstIncSign = ldmaCfgSrcIncSignNeg;
+		break;
+	case DMA_ADDR_ADJ_NO_CHANGE:
+		xfer_config->ldmaCfgDstIncSign = ldmaCfgSrcIncSignPos;
+		break;
+	default:
+		LOG_ERR("Addr Adjustement error. %d", config->head_block->source_addr_adj);
+		break;
+	}
+
+	switch (config->head_block->dest_addr_adj) {
+	case DMA_ADDR_ADJ_INCREMENT:
+		xfer_config->ldmaCfgSrcIncSign = ldmaCfgDstIncSignPos;
+		break;
+	case DMA_ADDR_ADJ_DECREMENT:
+		xfer_config->ldmaCfgSrcIncSign = ldmaCfgDstIncSignNeg;
+		break;
+	case DMA_ADDR_ADJ_NO_CHANGE:
+		xfer_config->ldmaCfgSrcIncSign = ldmaCfgDstIncSignPos;
+		break;
+	default:
+		break;
+	}
+
+	xfer_config->ldmaLoopCnt = 0;
+
+	ret = dma_silabs_configure_descriptor(config, data, chan_conf);
+	if (ret) {
+		return ret;
+	}
+
+	atomic_set_bit(data->dma_ctx.atomic, channel);
+
+	return ret;
+}
+
+static int dma_silabs_start(const struct device *dev, uint32_t channel)
+{
+
+	const struct dma_silabs_data *data = dev->data;
+	struct dma_silabs_channel *chan_conf = &data->dma_chan_table[channel];
+
+	if (channel > data->dma_ctx.dma_channels) {
+		return -EINVAL;
+	}
+
+	chan_conf->busy = true;
+
+	LDMA_StartTransfer(channel, &chan_conf->xfer_config, chan_conf->desc);
+
+	return 0;
+}
+
+static int dma_silabs_stop(const struct device *dev, uint32_t channel)
+{
+	const struct dma_silabs_data *data = dev->data;
+	struct dma_silabs_channel *chan = &data->dma_chan_table[channel];
+
+	if (channel > data->dma_ctx.dma_channels) {
+		return -EINVAL;
+	}
+
+	LDMA_StopTransfer(channel);
+
+	chan->busy = false;
+
+	/* Clear the interrupt flag. */
+	LDMA->IF_CLR = BIT(channel);
+
+	return 0;
+}
+
+static int dma_silabs_get_status(const struct device *dev, uint32_t channel,
+				 struct dma_status *status)
+{
+	const struct dma_silabs_data *data = dev->data;
+
+	if (channel > data->dma_ctx.dma_channels) {
+		return -EINVAL;
+	}
+
+	if (!atomic_test_bit(data->dma_ctx.atomic, channel)) {
+		return -EINVAL;
+	}
+
+	status->busy = data->dma_chan_table[channel].busy;
+	status->dir = data->dma_chan_table[channel].dir;
+
+	return 0;
+}
+
+static int dma_silabs_init(const struct device *dev)
+{
+	struct dma_silabs_config *config = (struct dma_silabs_config *)dev->config;
+	struct dma_silabs_data *data = (struct dma_silabs_data *)dev->data;
+	struct dma_silabs_channel *dma_channel;
+	LDMA_Init_t dmaInit = LDMA_INIT_DEFAULT;
+
+	if (!device_is_ready(config->clock_dev)) {
+		LOG_ERR("clock control device not ready");
+		return -ENODEV;
+	}
```

Comment:
I think `clocks` is a standard attribute managed by the Zephyr framework. It is probably useless to check it.

Reply:
Related to this, it's a bit surprising that the LDMA peripheral clock doesn't get enabled, I would have expected a call to `clock_control_on()` somewhere. But I notice that the HAL for some reason does the clock enable internally, which is atypical for emlib.